core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -122,7 +122,7 @@ public TypeCoercionImpl(RelDataTypeFactory typeFactory, SqlValidator validator)
       final SqlCall operand0 = ((SqlCall) query).operand(0);
       final SqlCall operand1 = ((SqlCall) query).operand(1);
       final boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType)
-          && rowTypeCoercion(scope, operand1, columnIndex, targetType);
+          || rowTypeCoercion(scope, operand1, columnIndex, targetType);
       // Update the nested SET operator node type.
```

Comment:
Thanks, instead of `||` or `&&` to chain each union branch type coercion, we should use a `for-loop` to ensure each branch are checked.

You can add a test case in the `TypeCoercionTest#testSetOperations`.

Reply:
@danny0405 Thanks, Danny!

Reply:
@Aaaaaaron  any update on the test in `TypeCoercionTest#testSetOperations`?

Reply:
@amaliujia didn't reproduce the case in that ut, will take a look later, thanks.